### PR TITLE
LDAP documentation

### PR DESCRIPTION
**Pull request by anusudarsan**
@amrutagokhale @ebd2 Please review b284f4c and ec27f83. I based this off FB/master so that we have the latest security doc changes (and hence the rest of the diff).  This wont be merged to our sprint-branch now.

### CODE REVIEW

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+coordinator validates these credentials using an external LDAP service.
+The LDAP implementation types supported by Presto are Active Directory
+and OpenLDAP.
+
+We recommend to always use Secure LDAP (LDAPS) in production. So before 
+configuring Presto for LDAP Authentication, make sure that you have
+TLS enabled on your LDAP Server. You need to import the LDAP Server's
+TLS Certificate to the default Java truststore of the Presto coordinator
+to secure TLS connection. You can use the `keytool` command to
+import this certificate ``ldap_server.crt`` to the truststore
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: does the name have to be ldap_server.crt?

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+Presto coordinator to use LDAP authentication and HTTPS. 
+
+ * :ref:`ldap_server`
+ * :ref:`server_java_keystore`
+
+config.properties
+~~~~~~~~~~~~~~~~~
+
+LDAP authentication is configured in the coordinator node's
+config.properties file. The entries that need to be added are listed below.
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: s/The entries/Sample entries or Example for entries

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    http-server.https.enabled=true
+    http-server.https.port=8443
+
+    http-server.https.keystore.path=/etc/presto_keystore.jks
+    http-server.https.keystore.key=keystore_password
+
+======================================================= ======================================================
+Property                                                Description
+======================================================= ======================================================
+``authentication.ldap.enabled``                         Enable LDAP Authentication for the Presto coordinator.
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: It might be a good idea to specify what are the default values for these properties.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+``http-server.https.enabled``                           Enables HTTPS access for the Presto coordinator.
+                                                        Should be set to ``true``.
+``http-server.https.port``                              HTTPS server port.
+``http-server.https.keystore.path``                     The location of the Java Keystore file that will be
+                                                        used to secure TLS.
+``http-server.https.keystore.key``                      The password for the keystore. This must match the
+                                                        password you specified when creating the keystore.
+======================================================= ======================================================
+
+Group Based Authorization Support
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Is it Authorization or Authentication?

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+validated to determine if that user belongs to a specified group. This feature
+is optional and is enabled if you set ``authentication.ldap.group-dn`` property.
+Only users belonging to the group with distinguished name
+``authentication.ldap.group-dn`` are authenticated successfully.
+
+.. note::
+
+    Presto does not yet support providing multiple groups for 
+    ``authentication.ldap.group-dn``. Nested group authorization, where you
+    want to authorize an user who belongs to a group which is a member of
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: s/an user/a user

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+TLS Configuration
+~~~~~~~~~~~~~~~~~
+
+Access to the Presto coordinator should be through https when using LDAP
+authentication. The Presto CLI can use either a :ref:`Java Keystore
+<server_java_keystore>` file or :ref:`Java Truststore <cli_java_truststore>`
+for its TLS configuration. 
+
+If using Keystore file it can be copied to the client machine and used
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: s/Keystore/keystore?

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+for its configuration. If using Truststore you can either use default
+Java truststores or create a custom truststore on the CLI. We do not
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: s/Truststore/truststore?

File: presto-docs/src/main/sphinx/security/tls.rst [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+--------------------------
+
+Access to the Presto coordinator must be through HTTPS when using Kerberos
+and LDAP authentication. The Presto coordinator uses a :ref:`Java Keystore
+<server_java_keystore>` file for its TLS configuration. These keys are
+generated using :command:`keytool` and stored in a Java Keystore file for the
+Presto coordinator.
+
+The alias in the :command:`keytool` command line should match the principal that the
+Presto coordinator will use.
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/tls.rst]: will use for Kerberos authentication

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                                password you specified when creating the keystore.
+``--truststore-path``           The location of the Java Truststore file that will be used
+                                to secure TLS.
+``--truststore-password``       The password for the Truststore. This must match the
+                                password you specified when creating the Truststore.
+``--user``                      The LDAP username.
+``--password``                  The password for the ``user``.
+=============================== =========================================================================
+
+Troubleshooting
```
Comment (akshatnair) [inline presto-docs/src/main/sphinx/security/ldap.rst]: This seems to be not required and present at the end of this diff as well

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 10]
```diff
@@ -0,0 +1,212 @@
+===================
+LDAP Authentication
+===================
+
+Presto can be configured to enable frontend LDAP authentication over
+HTTPS for clients, such as the :ref:`cli_ldap`, or the JDBC and ODBC
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst:6]: Should we mention about the JDBC/ODBC drivers here, given that the open source drives don't have LDAP support?

Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst:6]: hmm I copied this from the kerberos documentation, and there we add drivers too.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 1 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: no I was referring to the example.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 2 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: done

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 3 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: added defaults for boolean properties. Everything else defaults to empty string, so leaving as-is.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 4 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: its authorization but based on LDAP. Renamed to "Authorization based on LDAP group membership"

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 9 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: I was trying to link the other one from here.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 6 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: done

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 7 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: done

File: presto-docs/src/main/sphinx/security/tls.rst (hunk 8 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/tls.rst]: done

**Comment by anusudarsan** [comment]
addressed comments

**Comment by amrutagokhale** [comment]
Do you intend to add two sections -- one for the CLI, say `CLI LDAP Authentication` and the other for the Presto coordinator, say `Coordinator LDAP Authentication`, similar to the Kerberos documentation?

**Comment by anusudarsan** [comment]
I kept it under one page since it was not as a big as the kerberos doc. I can split it up if you think that will look better.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+Environment Configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _ldap_server:
+
+LDAP Server
+~~~~~~~~~~~
+
+The Presto clients send a user name and password to the coordinator and
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: s/Presto clients send/Presto client sends/

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 12]
```diff
@@ -0,0 +1,213 @@
+===================
+LDAP Authentication
+===================
+
+Presto can be configured to enable frontend LDAP authentication over
+HTTPS for clients, such as the :ref:`cli_ldap`, or the JDBC and ODBC
+drivers. This involves user and password validation backed by LDAP.
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: s/user/user name/

**Comment by akshatnair** [comment]
lgtm for me

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 12 above)
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: How about something like this instead, which mentions that we only support "simple" LDAP authentication mechanism:

At present, only the <i>simple</i> LDAP authentication mechanism involving user name and password is supported.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+.. _ldap_server:
+
+LDAP Server
+~~~~~~~~~~~
+
+The Presto clients send a user name and password to the coordinator and
+coordinator validates these credentials using an external LDAP service.
+The LDAP implementation types supported by Presto are Active Directory
+and OpenLDAP.
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: s/LDAP implementation/LDAP server implementation. Or you could rewrite it as:
You can configure Presto coordinator to communicate with one of the following two types of LDAP server implementations:
- Active Directory
- OpenLDAP

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: You can restructure this section a bit. Put this under the header `Enable LDAP over TLS` and then you can itemize the content by saying:-
We recommend to always use secure LDAP (LDAPS) in production. Follow these steps to enable TLS for LDAP authentication:
1) Enable TLS on LDAP server (by doing x, y and z) 
2) Import the LDAP server's TLS certificate to the ......
3) Enforce use of HTTPS for the communication with the Presto coordinator. You can achieve this by ....

**Comment by amrutagokhale** [comment]
Nitpick: Please use lowercase letters for words that occur in the middle of the sentences and do not represent expansion of acronyms. For ex., Authentication -> authentication, Certificate -> certificate, Secure -> secure, Server -> server etc.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 12 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: done.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 13 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: done

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+Environment Configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. _ldap_server:
+
+LDAP Server
+~~~~~~~~~~~
+
+The Presto client sends a user name and password to the coordinator and
+coordinator validates these credentials using an external LDAP service.
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Since this sentence doesn't describe configuration per say, perhaps move this above, after say, "At present only simple LDAP authentication mechanism involving user name and password is supported.".

**Comment by anusudarsan** [comment]
@amrutagokhale addressed comments

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+Only users belonging to the group with distinguished name
+``authentication.ldap.group-dn`` are authenticated successfully.
+
+.. note::
+
+    Presto does not yet support providing multiple groups for 
+    ``authentication.ldap.group-dn``. Nested group authorization, where you
+    want to authorize a user who belongs to a group which is a member of
+    ``authentication.ldap.group-dn`` is also not supported.
+
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst:166]: Add sample values for the following properties as well.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+                                                        used to secure TLS.
+``http-server.https.keystore.key``                      The password for the keystore. This must match the
+                                                        password you specified when creating the keystore.
+======================================================= ======================================================
+
+Authorization based on LDAP Group Membership
+********************************************
+
+Once the user is successfully authenticated via LDAPS, the user can be 
+validated to determine if that user belongs to a specified group. This feature
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Perhaps rephrase this as: "You can further restrict the set of users allowed to connect to Presto coordinator based on their group memberships."

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+is optional and is enabled if you set ``authentication.ldap.group-dn`` property.
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: -> This optional feature is enabled if you set ....

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 16 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst:166]: its already under config.properties header above, do you want me to move those over here? I put it there because we talk about base-dn in the other properties table.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 17 above)
Comment (anusudarsan) [inline presto-docs/src/main/sphinx/security/ldap.rst]: done.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 19]
```diff
@@ -0,0 +1,239 @@
...
+the coordinator.
+
+Presto Coordinator Node Configuration
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You must make the following changes to the environment prior to configuring the
+Presto coordinator to use LDAP authentication and HTTPS. 
+
+ * :ref:`ldap_server`
+ * :ref:`server_java_keystore`
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst:59]: I believe you can move this entire section titled `Presto Coordinator Node Configuration` above, instead of back referencing it. The text starting with "You need to import the LDAP server's TLS certificate to the default Java truststore of the Presto coordinator" can be put under a subsection called `TLS Configuration on Presto Coordinator`.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 20]
```diff
@@ -0,0 +1,239 @@
...
+
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst:60]: You may want to introduce the text below with a line such as "You need to make following changes to the Presto configuration files". Otherwise, the text below seems disconnected.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+********************************************
+
+You can further restrict the set of users allowed to connect to the Presto
+coordinator based on their group memberships. This optional feature is enabled
+if you set ``authentication.ldap.group-dn`` property.
+Only users belonging to the group with distinguished name
+``authentication.ldap.group-dn`` are authenticated successfully.
+
+In addition to the basic LDAP authentication properties, you need group-membership specific
+properties in ``config.properties``. Sample entries for both Active Directory and OpenLDAP are
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Is it that the same entries will work for both, AD and OpenLDAP? Isn't `object-class` different between them?

Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Also, if the entries are same, I think it's better not to mention AD and OpenLDAP. Mentioning them gives an impression that you're going to describe entries for each one of them separately.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 22]
```diff
@@ -0,0 +1,239 @@
...
+
+TLS Configuration
+~~~~~~~~~~~~~~~~~
+
+Access to the Presto coordinator should be through HTTPS when using LDAP
+authentication. The Presto CLI can use either a :ref:`Java Keystore
+<server_java_keystore>` file or :ref:`Java Truststore <cli_java_truststore>`
+for its TLS configuration. 
+
+If using keystore file it can be copied to the client machine and used
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: -> If you are using keystore file, you can copy it to the client machine.

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+for its configuration. If using truststore you can either use default
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Not sure what you mean by "for its configuration" ? Also, add a comma to make it look like: "If you are using truststore, you can".

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+of the following types of LDAP server implementations:
+
+    * Active Directory
+    * OpenLDAP
+
+Enable TLS
+**********
+
+We recommend to always use Secure LDAP (LDAPS) in production. So before 
+configuring Presto for LDAP authentication, make sure that you have
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: "So before configuring Presto for LDAP authentication," -> "In order to use secure LDAP,"

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 19 above)
Comment (anusudarsan) [reply presto-docs/src/main/sphinx/security/ldap.rst:59]: > I believe you can move this entire section titled Presto Coordinator Node Configuration above, instead of back referencing it.

I was following the same format as Kerberos doc, where the Env config section comes first.

File: presto-docs/src/main/sphinx/security/ldap.rst (hunk 21 above)
Comment (anusudarsan) [reply presto-docs/src/main/sphinx/security/ldap.rst]: added separate samples for both.

**Comment by anusudarsan** [comment]
@amrutagokhale addressed comments

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 25]
```diff
@@ -0,0 +1,249 @@
...
+********************************************
+
+You can further restrict the set of users allowed to connect to the Presto
+coordinator based on their group memberships. This optional feature is enabled
+if you set ``authentication.ldap.group-dn`` property.
+Only users belonging to the group with distinguished name
+``authentication.ldap.group-dn`` are authenticated successfully.
+
+In addition to the basic LDAP authentication properties, you need group-membership specific
+properties in ``config.properties``. Sample entries are listed below.
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: -> Sample entires that need to be added are listed below.

Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Or should you rather say "Properties that need to be added, along with their sample values, are listed below." (this applies to everywhere else too wherever this sentence "Sample entries... are listed below" appears).

File: presto-docs/src/main/sphinx/security/ldap.rst [hunk 26]
```diff
@@ -0,0 +1,249 @@
...
+``authentication.ldap.server-type``                     The type of LDAP server implementation.
+                                                        Set to ``active_directory`` or ``openLDAP``.                                                        
+``authentication.ldap.ad-domain``                       The domain name of the Active Directory server.
+                                                        This property is required only if 
+                                                        ``authentication.ldap.server-type`` is set to
+                                                        ``Active_Directory``. Eg: ``corp.domain.com``.
+``authentication.ldap.base-dn``                         The base LDAP distinguished name for the user who
+                                                        who tries to connect to the server. This property is 
+                                                        required if ``authentication.ldap.server-type`` is set
+                                                        to ``openLDAP``.
```
Comment (amrutagokhale) [inline presto-docs/src/main/sphinx/security/ldap.rst]: Should we rather add this clause too: "or when specifying LDAP group membership properties." 
I know I had earlier said that we can remove this, but I am now thinking that the current text doesn't give the complete information.

**Comment by amrutagokhale** [comment]
Few minor comments. Looks good otherwise.

**Comment by anusudarsan** [comment]
addressed comments
